### ISSUE TITLE
[feature request] Include libomp support (macOS)

### ISSUE THREAD

**Issue by rien333** [pytorch/pytorch#6328]
The standard way to link OpenMP while using Apple Clang on macOS seems to be by using [libomp](https://openmp.llvm.org/), which is a bottle available in [brew](https://github.com/Homebrew/homebrew-core/blob/master/Formula/libomp.rb). Apple doesn't seem to include libomp in their version of the compiler, but this bottle fixes that. It seems as though CMake does not pick up  libomp however, even when following the instructions in the formula. It could be an error on my side however. This [guide](https://iscinumpy.gitlab.io/post/omp-on-high-sierra/) might also be of interest.

**Comment by soumith** [comment]
this is a good guide. However, I think the latest clang/llvm actually now supports OpenMP officially. 

> "With the release of Clang 3.8.0, OpenMP 3.1 support is enabled in Clang by default, and the OpenMP runtime is therefore built as a normal part of the Clang build, and distributed with the binary distributions.You do not, therefore, need explicitly to check out this code, or build it out of tree; a normal Clang check out and build will automatically include building these runtime libraries."

**Comment by rien333** [comment]
Yes, but the official build documentation uses Apple clang, which doesn't bundle OpenMP. So either the documentation needs to reflect that say a non Apple clang from brew needs to be used, or the build process should use be updated to be able to use the libomp library together with Apple clang.

**Comment by rien333** [comment]
@soumith  What do you recommend? Just using a non-apple clang? I think you closed this issue a bit too early, as Apple clang clearly does not support OpenMP (see also this [reddit thread](https://www.reddit.com/r/MacOS/comments/71hz1n/does_apple_bundled_clang_have_openmp_support_in/)):
```shell
$ clang --version # Newest command line tools for 10.13.4 install
Apple LLVM version 9.1.0 (clang-902.0.39.1)
$ clang -fopenmp test.c -o test
clang: error: unsupported option '-fopenmp'
```

**Comment by soumith** [comment]
officially, we cannot support non-apple tooling, it becomes really hard to give users a consistent experience without that. so, official openmp support is out.

If users figure out how to build their own tooling, which enable OpenMP support, I think that we can highlight that in the README somehow.

Saying that, we are moving our core internals away from OpenMP based parallelism and towards Intel TBB based. The transition will take quite some time, but that's the direction we're headed, and hence I'm less worried about the OpenMP story.

**Comment by rien333** [comment]
I see. I did end up figuring out how to enable it however. The easiest way might just be using this [cmake module](https://github.com/CLIUtils/cmake/blob/2acdd6dd3f61dddba3448ab231e4adcf274a3ec3/PatchOpenMPApple.cmake) , and including it in `aten/CmakeList.txt` above the OpenMP section. I have to say that it adds an pretty significant CPU perfomance boost.

**Comment by malfet** [comment]
OpenMP (libomp) has been supported on macOS for a long time — `torch.backends.openmp.is_available()` is `True` on the released macOS build:

```
$ python -c "import torch; print(torch.__version__, torch.backends.openmp.is_available(), torch.get_num_threads())"
2.13.0 True 6
```

Closing as implemented.
